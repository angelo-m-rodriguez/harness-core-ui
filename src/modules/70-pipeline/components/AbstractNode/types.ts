import type { IconName } from '@harness/uicore'
import type { CSSProperties } from 'react'

export interface ListenerHandle {
  deregister: () => any
  id: string
  listener: BaseListener
}

export interface NodeData {
  name: string
  icon: IconName
  selectedColour: string
  unSelectedColour: string
  selectedIconColour: string
  unSelectedIconColour: string
}

export type BaseListener = (event: any) => void

export const enum PipelineGraphType {
  STAGE_GRAPH = 'STAGE_GRAPH',
  STEP_GRAPH = 'STEP_GRAPH'
}
export interface PipelineGraphState {
  id: string
  identifier: string
  type: string
  name: string
  icon: IconName
  status?: string
  data: any
  nodeType?: string
  graphType?: PipelineGraphType
  children?: PipelineGraphState[]
  parentStepGroupId?: string
  readonly?: boolean
}
export interface NodeIds {
  startNode: string
  createNode: string
  endNode: string
}

export interface SVGPathRecord {
  [key: string]: string
}

export type NodeBank = Map<string, NodeDetails>
export interface NodeDetails {
  component: React.FC<BaseReactComponentProps>
  isDefault?: boolean
}

export interface NodeCollapsibleProps {
  /** parent element selector to listen resize event on */
  parentSelector: string
  /** percent child visible to collapse */
  percentageNodeVisible?: number
  /** margin from child bottom to start expanding */
  bottomMarginInPixels?: number
}

export enum NodeStatus {
  Loading = 'Loading',
  Success = 'Success',
  Failure = 'Failure'
}

export enum NodeType {
  Default = 'default-node',
  EmptyNode = 'empty-node',
  CreateNode = 'create-node',
  DiamondNode = 'default-diamond',
  StartNode = 'start-node',
  GroupNode = 'group-node',
  IconNode = 'icon-node',
  EndNode = 'end-node',
  StepGroupNode = 'StepGroup'
}

export interface NodeProps<T> {
  width: number
  height: number
  onUpdate?: (data: T) => void
  onChange?: (data: T) => void
}

export interface NodeInterface {
  identifier: string
  type: NodeType
  name: string
  defaultIcon: IconName
  secondaryIcon?: IconName
  selectedColour?: string
  unSelectedColour?: string
  selectedIconColour?: string
  unSelectedIconColour?: string
}

export type FireEventMethod = (arg0: {
  type: string
  target: EventTarget
  data: {
    allowAdd?: boolean
    entityType?: string
    identifier?: string
    parentIdentifier?: string
    prevNodeIdentifier?: string
    node?: any
    destination?: any
    nodesInfo?: NodeInfo[]
  }
}) => void

export interface NodeInfo {
  name: string
  icon: string
  identifier: string
  type: string
}

export type GetNodeMethod = (type?: string | undefined) => NodeDetails
export interface BaseReactComponentProps {
  getNode?: GetNodeMethod
  fireEvent?: FireEventMethod
  setSelectedNode?(identifier: string): void
  getDefaultNode?(): NodeDetails | null
  updateGraphLinks?(): void
  isFirstParallelNode?: boolean
  className?: string
  name?: string
  identifier?: string
  id: string
  icon?: string
  iconStyle?: CSSProperties
  readonly?: boolean
  onClick?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void
  isSelected?: boolean
  defaultSelected?: any
  parentIdentifier?: string
  isParallelNode?: boolean
  prevNodeIdentifier?: string
  nextNode?: PipelineGraphState
  allowAdd?: boolean
  prevNode?: PipelineGraphState
  type?: string
  selectedNodeId?: string
  nodesInfo?: NodeInfo[]
  width?: number
  height?: number
  data?: any
}
