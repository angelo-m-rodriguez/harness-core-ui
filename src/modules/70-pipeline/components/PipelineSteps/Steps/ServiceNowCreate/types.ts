/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
 */

import type { FormikProps } from 'formik'
import type { GetDataError } from 'restful-react'
import type { MultiSelectOption, MultiTypeInputType, SelectOption } from '@wings-software/uicore'
import type { InputSetData, StepViewType } from '@pipeline/components/AbstractSteps/Step'
import type {
  Failure,
  ResponseListServiceNowFieldNG,
  ResponseListServiceNowTicketTypeDTO,
  ServiceNowFieldNG,
  StepElementConfig,
  UseGetServiceNowIssueCreateMetadataProps,
  UseGetServiceNowTicketTypesProps
} from 'services/cd-ng'
import type { VariableMergeServiceResponse } from 'services/pipeline-ng'
import type { ServiceNowTicketTypeSelectOption } from '@pipeline/components/PipelineSteps/Steps/ServiceNowApproval/types'

export interface ServiceNowCreateFieldType {
  name: string
  value: string | number | SelectOption | MultiSelectOption[]
}

export interface ServiceNowFieldNGWithValue extends ServiceNowFieldNG {
  value: string | number | SelectOption | MultiSelectOption[]
}

export interface ServiceNowCreateData extends StepElementConfig {
  spec: {
    connectorRef: string | SelectOption
    ticketType: string | ServiceNowTicketTypeSelectOption
    description?: string
    selectedFields?: ServiceNowFieldNGWithValue[]
    delegateSelectors?: string[]
  }
}

export interface ServiceNowCreateVariableListModeProps {
  variablesData: ServiceNowCreateData
  metadataMap: Required<VariableMergeServiceResponse>['metadataMap']
}

export interface ServiceNowCreateStepModeProps {
  stepViewType: StepViewType
  initialValues: ServiceNowCreateData
  onUpdate?: (data: ServiceNowCreateData) => void
  onChange?: (data: ServiceNowCreateData) => void
  allowableTypes: MultiTypeInputType[]
  isNewStep?: boolean
  readonly?: boolean
}

export interface ServiceNowCreateFormContentInterface {
  formik: FormikProps<ServiceNowCreateData>
  isNewStep?: boolean
  readonly?: boolean
  stepViewType: StepViewType
  allowableTypes: MultiTypeInputType[]
  refetchServiceNowTicketTypes: (props: UseGetServiceNowTicketTypesProps) => Promise<void>
  serviceNowTicketTypesFetchError?: GetDataError<Failure | Error> | null
  fetchingServiceNowTicketTypes: boolean
  serviceNowTicketTypesResponse: ResponseListServiceNowTicketTypeDTO | null
  refetchServiceNowMetadata: (props: UseGetServiceNowIssueCreateMetadataProps) => Promise<void>
  serviceNowMetadataFetchError?: GetDataError<Failure | Error> | null
  fetchingServiceNowMetadata: boolean
  serviceNowMetadataResponse: ResponseListServiceNowFieldNG | null
}

export enum ServiceNowCreateFormFieldSelector {
  FIXED = 'FIXED',
  EXPRESSION = 'EXPRESSION'
}

export interface ServiceNowFieldSelectorProps {
  fields: ServiceNowFieldNG[]
  selectedFields: ServiceNowFieldNG[]
  addSelectedFields: (selectedFields: ServiceNowFieldNG[]) => void
  onCancel: () => void
}

export interface ServiceNowCreateDeploymentModeProps {
  stepViewType: StepViewType
  initialValues: ServiceNowCreateData
  onUpdate?: (data: ServiceNowCreateData) => void
  allowableTypes: MultiTypeInputType[]
  inputSetData?: InputSetData<ServiceNowCreateData>
  formik?: any
}

export interface ServiceNowCreateDeploymentModeFormContentInterface extends ServiceNowCreateDeploymentModeProps {
  refetchServiceNowTicketTypes: (props: UseGetServiceNowTicketTypesProps) => Promise<void>
  serviceNowTicketTypesFetchError?: GetDataError<Failure | Error> | null
  fetchingServiceNowTicketTypes: boolean
  serviceNowTicketTypesResponse: ResponseListServiceNowTicketTypeDTO | null
  refetchServiceNowMetadata: (props: UseGetServiceNowIssueCreateMetadataProps) => Promise<void>
  serviceNowMetadataFetchError?: GetDataError<Failure | Error> | null
  fetchingServiceNowMetadata: boolean
  serviceNowMetadataResponse: ResponseListServiceNowFieldNG | null
}
