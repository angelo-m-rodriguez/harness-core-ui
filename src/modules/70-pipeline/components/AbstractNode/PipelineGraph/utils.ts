import type { RefObject } from 'react'
import type { StageElementWrapperConfig } from 'services/cd-ng'
const INITIAL_ZOOM_LEVEL = 1
const ZOOM_INC_DEC_LEVEL = 0.1
const getFinalSVGArrowPath = (parentId: string, id1 = '', id2 = '', isParallelNode = false): string => {
  if (!parentId) {
    return ''
  }
  const node1 = getComputedPosition(parentId, id1)
  const node2 = getComputedPosition(parentId, id2)
  if (!node1 || !node2) {
    return ''
  }

  const node1VerticalMid = node1.top + node1.height / 2
  const node2VerticalMid = node2.top + node2.height / 2

  const startPoint = `${node1.right},${node1VerticalMid}`
  const horizontalMid = Math.abs((node1.right + node2.left) / 2)
  const endPoint = `${node2.left},${node2VerticalMid}`

  if (node2.y < node1.y) {
    //  child node is at top
    const curveLeftToTop = `Q${horizontalMid},${node1VerticalMid} ${horizontalMid},${node1VerticalMid - 20}`
    const curveBottomToRight = `Q${horizontalMid},${node2VerticalMid} ${horizontalMid + 20},${node2VerticalMid}`
    return `M${startPoint} L${horizontalMid - 20},${node1VerticalMid} ${curveLeftToTop} 
    L${horizontalMid},${node2VerticalMid + 20} ${curveBottomToRight} L${endPoint}`
  } else if (node1.y === node2.y) {
    // both nodes are at same level vertically
    return `M${startPoint}  L${endPoint} `
  } else {
    //  child node is at bottom
    const curveLeftToBottom = `Q${horizontalMid},${node1VerticalMid} ${horizontalMid},${node1VerticalMid + 20}`

    const curveTopToRight = `Q${horizontalMid},${node2VerticalMid} ${horizontalMid + 20},${node2VerticalMid}`

    if (isParallelNode) {
      const updatedStart = node1.left - 45 // new start point
      const parallelLinkStart = `${
        updatedStart // half of link length
      },${node1VerticalMid}`

      const curveLBparallel = `Q${updatedStart + 20},${node1VerticalMid} ${updatedStart + 20},${node1VerticalMid + 20} `
      const curveTRparallel = `Q${updatedStart + 20},${node2VerticalMid} ${updatedStart + 40},${node2VerticalMid}`

      return `M${parallelLinkStart} 
      ${curveLBparallel} 
      L${updatedStart + 20},${node2VerticalMid - 20} 
      ${curveTRparallel} 
      L${node2.left},${node2VerticalMid} 
      M${node2.right},${node2VerticalMid} 
      L${node2.right + 10},${node2VerticalMid}
      Q${node2.right + 25},${node2VerticalMid} ${node2.right + 25},${node2VerticalMid - 20}      
      L${node2.right + 25},${node1VerticalMid + 20}
      Q${node2.right + 25},${node1VerticalMid} ${node2.right + 40},${node1VerticalMid}
      `
    }
    return `M${startPoint} L${horizontalMid - 20},${node1VerticalMid} ${curveLeftToBottom} 
    L${horizontalMid},${node2VerticalMid - 20} ${curveTopToRight} L${endPoint}`
  }
}

const getComputedPosition = (parentId: string, childId: string): DOMRect => {
  const parentPos = document.getElementById(parentId)?.getBoundingClientRect() as DOMRect
  const childPos = document.getElementById(childId)?.getBoundingClientRect() as DOMRect
  const updatedTop = childPos.top - parentPos.top
  const updatedLeft = childPos.left - parentPos.left
  const updatedRight = updatedLeft + childPos.width
  const updatedBottom = updatedTop + childPos.height
  const updatedPos: DOMRect = {
    ...childPos,
    left: updatedLeft,
    top: updatedTop,
    right: updatedRight,
    bottom: updatedBottom,
    width: childPos.width,
    height: childPos.height,
    x: childPos.x,
    y: childPos.y
  }
  return updatedPos
}

const setupDragEventListeners = (canvasRef: RefObject<HTMLDivElement>): (() => void) => {
  let offset = [0, 0]
  const divOverlay = (canvasRef.current as HTMLDivElement).parentElement as HTMLDivElement
  const canvas = canvasRef.current as HTMLDivElement
  let isDown = false

  const onMouseMove = (e: any): void => {
    e.preventDefault()
    if (isDown) {
      canvas.style.left = e.clientX + offset[0] + 'px'
      canvas.style.top = e.clientY + offset[1] + 'px'
    }
  }

  const onMouseDown = (e: any): void => {
    isDown = true
    offset = [canvas.offsetLeft - e.clientX, canvas.offsetTop - e.clientY]
  }
  const onMouseUp = (): void => {
    isDown = false
  }

  divOverlay.addEventListener('mousedown', onMouseDown, true)
  document.addEventListener('mouseup', onMouseUp, true)
  document.addEventListener('mousemove', onMouseMove, true)
  return () => {
    divOverlay.removeEventListener('mousedown', onMouseDown)
    document.removeEventListener('mouseup', onMouseUp)
    document.removeEventListener('mousemove', onMouseMove)
  }
}

const getSVGLinksFromPipeline = (stages?: StageElementWrapperConfig[], resultArr: string[] = []): string[] => {
  let prevElement: StageElementWrapperConfig
  stages?.forEach(stage => {
    if (!prevElement) {
      prevElement = stage
      return
    }
    if (stage.parallel) {
      getParallelNodeLinks(stage.parallel, prevElement, resultArr)
    } else {
      resultArr.push(
        getFinalSVGArrowPath('tree-container', prevElement.stage?.identifier as string, stage.stage?.identifier, false)
      )
    }
    prevElement = stage
  })
  return resultArr
}
const getParallelNodeLinks = (
  stages: StageElementWrapperConfig[],
  prevStage: StageElementWrapperConfig,
  resultArr: string[] = []
): void => {
  const [firstStage, ...restStages] = stages
  resultArr.push(
    getFinalSVGArrowPath('tree-container', prevStage.stage?.identifier as string, firstStage.stage?.identifier, false)
  )
  restStages?.forEach(stage => {
    resultArr.push(
      getFinalSVGArrowPath('tree-container', firstStage.stage?.identifier as string, stage.stage?.identifier, true)
    )
  })
}

const getScaleToFitValue = (elm: HTMLElement, paddingFromBottom = 20): number => {
  return (
    1 /
    Math.max(
      elm.clientWidth / window.innerWidth,
      elm.clientHeight / (window.innerHeight - elm.offsetTop - paddingFromBottom)
    )
  )
}
export {
  getFinalSVGArrowPath,
  setupDragEventListeners,
  getSVGLinksFromPipeline,
  ZOOM_INC_DEC_LEVEL,
  INITIAL_ZOOM_LEVEL,
  getScaleToFitValue
}
