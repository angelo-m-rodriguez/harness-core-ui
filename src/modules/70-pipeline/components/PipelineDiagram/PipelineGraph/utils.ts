import { defaultTo } from 'lodash-es'
import type { PipelineGraphState } from '../types'

const shouldAttachRef = (
  intersectingIndex: number,
  isCurrentChildLast: boolean,
  indexRelativeToParent: number
): boolean => {
  return intersectingIndex === -1 // intersecting index is not set
    ? isCurrentChildLast // current child is not the last child in parallel
    : intersectingIndex === indexRelativeToParent // intersecting index is not equal to indexRelativeToParent (i.e index+1 as parent is 0)
}

const shouldRenderGroupNode = (attachRef: boolean, isCurrentChildLast: boolean): boolean => {
  return attachRef && !isCurrentChildLast // render group node is ref is attached and current node is not the last child
}

const isFirstNodeAGroupNode = (intersectingIndex: number, collapseOnIntersect = false, childLength = 0): boolean =>
  intersectingIndex === 0 && collapseOnIntersect && childLength > 0 // if intersecting index is 0 and collapseOnIntersect is enabled

const showChildNode = (indexRelativeToParent: number, intersectingIndex: number): boolean => {
  return !(indexRelativeToParent > intersectingIndex && intersectingIndex !== -1) // child index is greater than intersectingIndex and intersectingIndex is set to some value
}

const isNodeParallel = (node: PipelineGraphState): boolean => defaultTo(node?.children?.length, 0) > 0

export { shouldAttachRef, shouldRenderGroupNode, isFirstNodeAGroupNode, showChildNode, isNodeParallel }
