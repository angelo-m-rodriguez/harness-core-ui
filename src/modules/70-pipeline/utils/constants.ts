/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */
import type { IconName } from '@harness/uicore'
import type { ConnectorInfoDTO } from 'services/cd-ng'

import type { StepViewType } from '@pipeline/components/AbstractSteps/Step'

export const EXPRESSION_STRING = '<+expression>' // TODO: this needs to be exported from uicore for best use.

export enum CardVariant {
  Default = 'Default',
  Minimal = 'Minimal',
  MinimalWithActions = 'MinimalWithActions'
}

export const ConnectorRefWidth = {
  DeploymentFormView: 320,
  DeploymentForm: 320, // matches StepViewType key for getConnectorRefWidth function
  InputSetView: 310,
  DefaultView: 385,
  EditStageView: 366,
  EditStageViewInputSet: 308,
  RightBarView: 460
}

export const connectorTypes: { [key: string]: ConnectorInfoDTO['type'] } = {
  Aws: 'Aws',
  Gcp: 'Gcp'
}
export const stageTypeToIconMap: Record<string, IconName> = {
  Deployment: 'cd-main',
  CI: 'ci-main',
  Pipeline: 'pipeline',
  Custom: 'pipeline-custom',
  Approval: 'approval-stage-icon',
  FeatureFlag: 'cf-main'
}

export const getConnectorRefWidth = (viewType: StepViewType): number =>
  Object.entries(ConnectorRefWidth).find(key => key[0] === viewType)?.[1] || ConnectorRefWidth.DefaultView
