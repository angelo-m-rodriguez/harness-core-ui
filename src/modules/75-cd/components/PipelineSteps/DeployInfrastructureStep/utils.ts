/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { FormikProps } from 'formik'
import { isEmpty } from 'lodash-es'

import { getMultiTypeFromValue, MultiTypeInputType, SelectOption } from '@harness/uicore'
import type { EnvironmentResponseDTO, PipelineInfrastructure } from 'services/cd-ng'

import type { StepViewType } from '@pipeline/components/AbstractSteps/Step'

export interface PipelineInfrastructureV2 extends PipelineInfrastructure {
  environmentOrEnvGroupRef?: SelectOption
  environmentGroup?: any
  environmentRef2?: any
  infrastructureRef?: any
}

export interface DeployInfrastructureProps {
  initialValues: PipelineInfrastructureV2
  onUpdate?: (data: PipelineInfrastructureV2) => void
  readonly: boolean
  allowableTypes: MultiTypeInputType[]
  stepViewType?: StepViewType
  inputSetData?: {
    template?: PipelineInfrastructureV2
    path?: string
    readonly?: boolean
  }
}

export interface DeployInfrastructureState {
  isEdit: boolean
  isEnvironment: boolean
  formik?: FormikProps<PipelineInfrastructureV2>
  data?: EnvironmentResponseDTO
}

export function isEditEnvironment(data: PipelineInfrastructureV2): boolean {
  if (
    getMultiTypeFromValue(data.environmentRef) !== MultiTypeInputType.RUNTIME &&
    !isEmpty(data.environmentOrEnvGroupRef)
  ) {
    return true
  }
  return false
}

export function isEditInfrastructure(data: any): boolean {
  if (
    getMultiTypeFromValue(data.infrastructureRef) !== MultiTypeInputType.RUNTIME &&
    !isEmpty(data.infrastructureRef)
  ) {
    return true
  } else if (data.infrastructure && !isEmpty(data.infrastructure.identifier)) {
    return true
  }
  return false
}
