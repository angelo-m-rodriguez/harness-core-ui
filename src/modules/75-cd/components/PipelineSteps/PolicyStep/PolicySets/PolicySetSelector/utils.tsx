/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { FontVariation, Layout, Text } from '@harness/uicore'
import { useStrings } from 'framework/strings'

export const PolicySetSelectorDialogTitle = (): JSX.Element => {
  const { getString } = useStrings()
  return (
    <Layout.Horizontal
      spacing="xsmall"
      padding={{ top: 'xxlarge', left: 'medium', right: 'large', bottom: 'medium' }}
      flex={{ justifyContent: 'space-between' }}
    >
      <Layout.Vertical>
        <Text font={{ variation: FontVariation.H3 }}>{getString('common.policiesSets.selectPolicySet')}</Text>
      </Layout.Vertical>
    </Layout.Horizontal>
  )
}
