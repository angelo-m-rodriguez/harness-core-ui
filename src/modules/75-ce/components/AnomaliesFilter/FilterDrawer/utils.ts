/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { MultiSelectOption } from '@harness/uicore'
import { omit } from 'lodash-es'
import type { AnomalyFilterProperties } from 'services/ce'
import type { AnomaliesFilterFormType } from './FilterDrawer'

const getMultiSelectOptions = (values: string[]): MultiSelectOption[] =>
  values.map(item => ({ label: item, value: item }))

const getValueFromOption = (values: MultiSelectOption[]): string[] => values.map(item => item.value.toString())

export const getAnomalyFormValuesFromFilterProperties = (
  filterProperties: AnomalyFilterProperties
): AnomaliesFilterFormType => {
  const formValues: AnomaliesFilterFormType = {}

  Object.entries(omit(filterProperties, 'filterType')).forEach(([key, value]) => {
    if (value) {
      if (key === 'minActualAmount' || key === 'minAnomalousSpend') {
        formValues[key] = +value
      } else {
        ;(formValues as any)[key] = getMultiSelectOptions(value as string[])
      }
    }
  })

  return formValues
}

export const getAnomalyFilterPropertiesFromForm = (formData: AnomaliesFilterFormType): AnomalyFilterProperties => {
  const filterProperties: AnomalyFilterProperties = { filterType: 'Anomaly' }

  Object.entries(formData).forEach(([key, value]) => {
    if (key === 'minActualAmount' || key === 'minAnomalousSpend') {
      filterProperties[key] = +value
    } else {
      ;(filterProperties as any)[key] = getValueFromOption(value)
    }
  })

  return filterProperties
}
