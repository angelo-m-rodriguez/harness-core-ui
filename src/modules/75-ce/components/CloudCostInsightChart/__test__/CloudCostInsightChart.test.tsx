/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { Provider } from 'urql'
import { fromValue } from 'wonka'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { CCM_CHART_TYPES } from '@ce/constants'
import { QlceViewTimeGroupType } from 'services/ce/services'
import { CCM_PAGE_TYPE } from '@ce/types'
import CloudCostInsightChart from '../CloudCostInsightChart'

const params = {
  accountId: 'TEST_ACC',
  perspetiveId: 'perspectiveId',
  perspectiveName: 'sample perspective'
}

describe('Test cases for cloud cost insight charts', () => {
  test('Should be able to render the chart component', async () => {
    const responseState = {
      executeQuery: () => {
        return fromValue({})
      }
    }

    const { container } = render(
      <TestWrapper pathParams={params}>
        <Provider value={responseState as any}>
          <CloudCostInsightChart
            fetching={false}
            data={null}
            columnSequence={[]}
            chartType={CCM_CHART_TYPES.LINE}
            aggregation={QlceViewTimeGroupType.Day}
            xAxisPointCount={0}
          />
        </Provider>
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })

  test('Should show loading in case data is not present', () => {
    const responseState = {
      executeQuery: () => {
        return fromValue({})
      }
    }

    const { container } = render(
      <TestWrapper pathParams={params}>
        <Provider value={responseState as any}>
          <CloudCostInsightChart
            fetching={true}
            data={null}
            columnSequence={[]}
            chartType={CCM_CHART_TYPES.LINE}
            aggregation={QlceViewTimeGroupType.Day}
            xAxisPointCount={0}
            pageType={CCM_PAGE_TYPE.Node}
          />
        </Provider>
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()

    const spinner = container.querySelector('[data-testid="loader"]')
    expect(spinner).toBeTruthy()
  })
})
