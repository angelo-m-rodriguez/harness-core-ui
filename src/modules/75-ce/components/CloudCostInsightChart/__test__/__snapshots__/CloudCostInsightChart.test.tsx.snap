// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test cases for cloud cost insight charts Should be able to render the chart component 1`] = `
<div>
  <div />
</div>
`;

exports[`Test cases for cloud cost insight charts Should show loading in case data is not present 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main chartMainContainer loadingContainer"
    data-testid="loader"
  >
    <span
      class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-blue500"
      data-icon="spinner"
    >
      <svg
        fill="none"
        height="28"
        viewBox="0 0 36 36"
        width="28"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 2.084a15.915 15.915 0 010 31.831 15.915 15.915 0 010-31.83"
          stroke="currentColor"
          stroke-dasharray="75,100"
          stroke-linecap="round"
          stroke-width="2.8"
          style="animation: 1s linear infinite spinner-rotation; transform-origin: 50% 50%; transform-box: fill-box;"
        />
      </svg>
    </span>
  </div>
</div>
`;
